### PR TITLE
Insert into referenced elements using $first, $last, index

### LINKED ISSUE #94: Catmandu-Wildcard-Support: $first, $last
It seems that `$first` and `$last` is not supported yet. (Catmandu is not consistent if `$first`, `$last` or `$start`, `$end` is the right wildcard. See: https://github.com/LibreCat/Catmandu/issues/387) (Update 20.12.: `$first` and `$last` are the correct ones: [See here ](https://github.com/LibreCat/Catmandu/issues/387#issuecomment-997658004))

[Example here](https://metafacture.org/playground/?data=%7B%0A++++%22arrayOfObjectimals%22%3A+%5B%7B%22name%22%3A+%22cat%22%7D%2C%7B%22name%22%3A+%22dog%22%7D%2C%7B%22name%22%3A+%22fox%22%7D%5D%0A%7D&flux=PG_DATA%0A%7C+as-records%0A%7C+decode-json%0A%7C+fix%0A%7C+encode-json%28prettyPrinting%3D%22true%22%29%0A%7C+print%0A%3B&fix=add_field%28%22arrayOfObjectimals%5B%5D.%24first.kind%22%2C%22angry%22%29%0Aadd_field%28%22arrayOfObjectimals%5B%5D.%24last.kind%22%2C%22nice%22%29&morph=&active-editor=fix)

Output should be:
```
{
    "arrayOfObjectimals": [
                               {"name": "cat", "kind": "angry"},
                               {"name": "dog"},
                               {"name": "fox", "kind": "nice"}
                          ]
}
```

### PR DESCRIPTION
**Pull request by fsteeg**
Will resolve #94

### CODE REVIEW

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 1]
```diff
@@ -475,22 +476,41 @@ private void insert(final InsertMode mode, final String[] fields, final String n
...
+            }
+        }
+
+        private void processDefault(final InsertMode mode, final String[] fields, final String newValue) {
+            if (isNumber(fields[0])) {
+                // TODO: WDCD? insert at the given index? also descend into the array?
+                if (fields.length == 1) {
+                    add(new Value(newValue));
+                }
+                if (fields.length > 1) {
```
Comment (blackwinter) [inline metafix/src/main/java/org/metafacture/metafix/Value.java:499]:
```suggestion
                else if (fields.length > 1) {
```

**Review by blackwinter — APPROVED** [review]
Feels like lots of added complexity, accompanied by unknown side effects :( One can only hope that it doesn't get out of hand in the future.

For now, I'm getting a bit uneasy about the divergence of object traversal between `find()`/`insert()` and `transformFields()`. I should probably work on that next.

Maybe you want to rethink your decision to reserve `$` field names, but otherwise :+1:

**Comment by fsteeg** [comment]
Thanks for the review, I'll integrate your suggestions before merging.

> Feels like lots of added complexity, accompanied by unknown side effects :( One can only hope that it doesn't get out of hand in the future.

I very much have the feeling that once we get to a certain functional stage, we'll be able to consolidate a lot of the complexity.

> For now, I'm getting a bit uneasy about the divergence of object traversal between `find()`/`insert()` and `transformFields()`. I should probably work on that next.

Right, this is also where I'm currently at in #95, where certain things only work in a bind if we first do e.g. an `upcase` (which calls `transformFields`).

File: metafix/src/main/java/org/metafacture/metafix/Value.java [hunk 2]
```diff
@@ -696,6 +714,41 @@ private Value insert(final InsertMode mode, final String[] fields, final String
...
+
+        private InsertMode insertMode(final InsertMode mode, final String field, final String[] tail) {
+            // if the field is marked as array, this hash should be smth. like { 1=a, 2=b }
+            final boolean isIndexedArray = field.endsWith(Metafix.ARRAY_MARKER);
+            final boolean nextIsRef = tail.length > 0 && (tail[0].startsWith(FIRST_FIELD) || tail[0].startsWith(LAST_FIELD) || isNumber(tail[0]));
+            return isIndexedArray && !nextIsRef ? InsertMode.INDEXED : mode;
+        }
+
+        private boolean isRef(final String field) {
+            return field.startsWith("$") || isNumber(field);
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply metafix/src/main/java/org/metafacture/metafix/Value.java]: I agree about the enum and started working on it. To keep the touched code here small, and to not sneak in the enum without review, I changed the condition to explicitly test all fields and added a TODO here. I will extract the enum separately.

**Pull request merged**